File: ios/Pacor/Utils/FastingColor.swift
import SwiftUI

enum FastingColor {
    private struct RGB {
        var r: Double
        var g: Double
        var b: Double
    }

    private struct ColorStop {
        let at: Double
        let color: RGB
    }

    private static let stops: [ColorStop] = [
        ColorStop(at: 0.0, color: RGB(r: 145, g: 148, b: 158)),
        ColorStop(at: 0.4, color: RGB(r: 220, g: 52, b: 52)),
        ColorStop(at: 0.8, color: RGB(r: 220, g: 52, b: 52)),
        ColorStop(at: 0.87, color: RGB(r: 255, g: 140, b: 150)),
        ColorStop(at: 0.94, color: RGB(r: 140, g: 220, b: 160)),
        ColorStop(at: 1.0, color: RGB(r: 60, g: 190, b: 100)),
    ]

    static func color(fromRatio ratio: Double) -> Color {
        let rgb = multiStopLerp(ratio: ratio)
        return Color(
            red: rgb.r / 255,
            green: rgb.g / 255,
            blue: rgb.b / 255
        )
    }

    static func gradientStops() -> [Gradient.Stop] {
        stops.map { stop in
            let rgb = stop.color
            return Gradient.Stop(
                color: Color(red: rgb.r / 255, green: rgb.g / 255, blue: rgb.b / 255),
                location: stop.at
            )
        }
    }

    private static func multiStopLerp(ratio: Double) -> RGB {
        let r = min(max(ratio, 0), 1)

        if r <= stops[0].at { return stops[0].color }
        if r >= stops[stops.count - 1].at { return stops[stops.count - 1].color }

        for index in 0..<(stops.count - 1) {
            let current = stops[index]
            let next = stops[index + 1]
            if r >= current.at && r <= next.at {
                let t = (r - current.at) / (next.at - current.at)
                return lerpRgb(a: current.color, b: next.color, t: t)
            }
        }

        return stops[stops.count - 1].color
    }

    private static func lerpRgb(a: RGB, b: RGB, t: Double) -> RGB {
        RGB(
            r: a.r + (b.r - a.r) * t,
            g: a.g + (b.g - a.g) * t,
            b: a.b + (b.b - a.b) * t
        )
    }
}

File: frontend/src/utils/fastingColor.ts
function lerp(a: number, b: number, t: number) {
  return a + (b - a) * t
}

function clamp01(value: number) {
  return Math.min(Math.max(value, 0), 1)
}

type Rgb = { r: number; g: number; b: number }

interface ColorStop {
  at: number
  color: Rgb
}

const STOPS: ColorStop[] = [
  { at: 0.0, color: { r: 145, g: 148, b: 158 } },
  { at: 0.4, color: { r: 220, g: 52, b: 52 } },
  { at: 0.8, color: { r: 220, g: 52, b: 52 } },
  { at: 0.87, color: { r: 255, g: 140, b: 150 } },
  { at: 0.94, color: { r: 140, g: 220, b: 160 } },
  { at: 1.0, color: { r: 60, g: 190, b: 100 } },
]

function lerpRgb(a: Rgb, b: Rgb, t: number): Rgb {
  return {
    r: lerp(a.r, b.r, t),
    g: lerp(a.g, b.g, t),
    b: lerp(a.b, b.b, t),
  }
}

function rgbToCss({ r, g, b }: Rgb): string {
  return `rgb(${Math.round(r)}, ${Math.round(g)}, ${Math.round(b)})`
}

function multiStopLerp(ratio: number): Rgb {
  const r = clamp01(ratio)

  if (r <= STOPS[0].at) return STOPS[0].color
  if (r >= STOPS[STOPS.length - 1].at) return STOPS[STOPS.length - 1].color

  for (let i = 0; i < STOPS.length - 1; i++) {
    const curr = STOPS[i]
    const next = STOPS[i + 1]
    if (r >= curr.at && r <= next.at) {
      const t = (r - curr.at) / (next.at - curr.at)
      return lerpRgb(curr.color, next.color, t)
    }
  }

  return STOPS[STOPS.length - 1].color
}

export function fastingColorFromRatio(ratio: number): string {
  return rgbToCss(multiStopLerp(ratio))
}

export function fastingGradientCss(): string {
  const stops = STOPS.map((stop) => `${rgbToCss(stop.color)} ${stop.at * 100}%`).join(', ')
  return `linear-gradient(to right, ${stops})`
}

export function fastingColor(elapsedMinutes: number, targetHours: number): string {
  const targetMinutes = Math.max(targetHours, 1) * 60
  return fastingColorFromRatio(elapsedMinutes / targetMinutes)
}
